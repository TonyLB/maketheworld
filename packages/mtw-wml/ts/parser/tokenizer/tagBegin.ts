import { Tokenizer, TokenBeginTagOpen } from "./baseClasses"

export const beginTagOpenTokenizer: Tokenizer<TokenBeginTagOpen> = (sourceStream) => {
    if (sourceStream.lookAhead('<')) {
        const startIdx = sourceStream.position
        let returnValue = sourceStream.consume(1)
        let tag = ''
        while(sourceStream.lookAhead(1).match(/[A-Za-z]/)) {
            tag = tag + sourceStream.consume(1)
        }
        returnValue = returnValue + tag
        if (returnValue) {
            return {
                type: 'BeginTagOpen',
                source: returnValue,
                tag: tag,
                startIdx,
                endIdx: sourceStream.position - 1
            }
        }
    }
    return undefined
}
