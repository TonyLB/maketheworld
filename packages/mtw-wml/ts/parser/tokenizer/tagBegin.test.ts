import SourceStream from './sourceStream'
import { beginTagOpenTokenizer } from './tagBegin'

describe('beginTagOpenTokenizer', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        jest.resetAllMocks()
    })

    it('should return undefined when no match', () => {
        const testStream = new SourceStream('Testing')
        expect(beginTagOpenTokenizer(testStream)).toBe(undefined)
    })
    it('should tokenize tag open', () => {
        const testStream = new SourceStream('<Room key=(ABC)>')
        expect(beginTagOpenTokenizer(testStream)).toMatchSnapshot()
    })
})