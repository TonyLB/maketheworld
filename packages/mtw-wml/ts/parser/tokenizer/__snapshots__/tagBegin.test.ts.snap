// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`beginTagOpenTokenizer should tokenize tag open 1`] = `
Object {
  "endIdx": 4,
  "source": "<Room",
  "startIdx": 0,
  "tag": "Room",
  "type": "BeginTagOpen",
}
`;
